src/common/redis_module/chain_module.h

```diff
@@ -0,0 +1,50 @@
+#ifndef RAY_CHAIN_MODULE_H_
+#define RAY_CHAIN_MODULE_H_
+
+#include <functional>
+
+#include "redismodule.h"
+
+// NOTE(zongheng): this duplicated declaration serves as forward-declaration
+// only.  The implementation is supposed to be linked in from credis.  In
+// principle, we can expose a header from credis and simple include that header.
+// This is left as future work.
+
+// Typical usage to make an existing redismodule command chain-compatible:
+//
+//   extern RedisChainModule module;
+//   int MyCmd_RedisModuleCmd(...) {
+//       return module.Mutate(..., NodeFunc, TailFunc);
+//   }
+//
+// See, for instance, ChainTableAdd_RedisCommand in ray_redis_module.cc.
+class RedisChainModule {
+ public:
+  // A function that runs on every node in the chain.  Type:
+  //   (context, argv, argc, (can be nullptr) mutated_key_str) -> int
+  //
+  // If the fourth arg is passed, NodeFunc must fill in the key being mutated.
```

Comment:
I'm a little confused by this comment. Can you be more specific about this optional argument, e.g., when it should be passed? Also, what does `NodeFunc` fill in the key with?